### PR TITLE
[RFC] Parsetree: represent punning and argument labels explicitly

### PR DESCRIPTION
**Pull request by nojb**
This PR is a follow-up to #8823, where it was discussed the possibility of representing puns explicitly in the Parsetree. This is in turn related to the representation of labelled arguments which today uses a type (`Asttypes.arg_label`) which allows many illegal constructions to be represented.

So far this PR contains a single commit with the proposed modifications to the Parsetree by which both constructions would be represented explicitly and accurately. The downside is that it is a rather invasive change to the Parsetree (I can only imagine the nightmares it will cause to the maintainers of the ppx infrastructure).

There is an intermediate version of this patch where we represent puns explictly but we leave `arg_label` alone, except that we add location information to the labels in question (today they have type `string`).

Let me know what you think. Thanks!

### CODE REVIEW

File: parsing/parsetree.mli [hunk 1]
```diff
@@ -220,9 +217,9 @@ and pattern_desc =
         (* `A             (None)
            `A P           (Some P)
          *)
-  | Ppat_record of (Longident.t loc * pattern) list * closed_flag
-        (* { l1=P1; ...; ln=Pn }     (flag = Closed)
-           { l1=P1; ...; ln=Pn; _}   (flag = Open)
+  | Ppat_record of (Longident.t loc * pattern option) list * closed_flag
+        (* { l1; ...; ln=Pn }        [(l1, None); ...; (ln, Some Pn)], Closed
+           { l1=P1; ...; ln=Pn; _}   [(l1, None); ...; (ln, Some Pn)], Open
```
Comment (lpw25) [inline parsing/parsetree.mli]: I think you want to drop the second `P1` as well

Comment (keleshev) [inline parsing/parsetree.mli]: Typo?

```suggestion
           { l1=P1; ...; ln=Pn; _}   [(l1, Some P1); ...; (ln, Some Pn)], Open
```

Comment (nojb) [reply parsing/parsetree.mli]: Yes, thanks.

Comment (nojb) [reply parsing/parsetree.mli]: Yes, thanks.

**Comment by garrigue** [comment]
While I like the idea of making labels "correct by typing", did you experiment with the consequences in the type checker ?
This should give you a first taste of how it will feel like using that.
I expect extensive changes in `type_application`. If it makes the code more readable, this is a net improvement.

**Pull request closed without merging**